k8s-deploy/pkg/cli/validation.go

```diff
@@ -0,0 +1,319 @@
+			return true
+		}
+	default:
+		panic(fmt.Sprintf("invalid type %v ", setVal.Type().Kind()))
+	}
+
+	return false
+}
+
+func compareTwoTrees(rootTemp, rootTest *TreeNode, testLines, skippedKeys []string) (errs []error) {
+	// recursively compare the two trees, walking through the nodes not skipped
+	valueTemp, valueTest := rootTemp.value, rootTest.value
+	typeTemp, typeTest := reflect.TypeOf(valueTemp), reflect.TypeOf(valueTest)
+	if typeTemp != typeTest {
+		route := strings.Join(rootTest.route, "/")
+		errs = append(errs, fmt.Errorf("Your yaml at '%s', line %d \n  '%s' may not match the type\n", route, rootTest.lineno, testLines[rootTest.lineno]))
+		return
+	}
+	switch valueTest := valueTest.(type) {
+	case KeyValue:
+		for k, v := range valueTest {
+			if Contains(k, skippedKeys) {
+				continue
+			}
+			if childTemp, ok := valueTemp.(KeyValue)[k]; !ok {
+				route := strings.Join(v.route, "/")
+				errs = append(errs, fmt.Errorf("Your yaml at '%s', line %d \n  '%s' may be redundant\n", route, v.lineno, testLines[v.lineno]))
+			} else {
+				errs = append(errs, compareTwoTrees(childTemp, v, testLines, skippedKeys)...)
+			}
+		}
+	case ListValue:
+		item := valueTemp.(ListValue)[0]
+		for _, v := range valueTest {
+			compareTwoTrees(item, v, testLines, skippedKeys)
+		}
+	default:
+	}
+	return
+}
+
+func (m *ValidationManager) Validate() error {
+	compareTwoTrees(m.templateTree.root, m.testTree.root, m.testTree.lines, m.skippedKeys)
+	return nil
+}
+
+func versionValid(chartVersion string, startVersion []int) (valid bool) {
+	// check if the chart version is valid
+	chartVersion = strings.TrimLeft(chartVersion, "v")
+	for i, v := range strings.Split(chartVersion, ".") {
+		if i >= len(startVersion) {
+			return
+		}
+		if v, err := strconv.Atoi(v); err != nil || v < startVersion[i] {
+			return
+		} else if v > startVersion[i] {
+			break
+		}
+	}
+	valid = true
+	return
+}
+
+func GetValueTemplateExample(chartName, chartVersion string) (value string, err error) {
+	// get the value template example from api
+	if !versionValid(chartVersion, []int{1, 9, 0}) {
+		err = errors.New("Yaml validation requires the chartVersion >= 1.9.0")
+		return
+	}
+
+	c := new(Chart)
+	req := &Request{
+		Type: "GET",
+		Path: fmt.Sprintf("chart/valueTemplateExample/%s/%s", chartName, chartVersion),
+		Body: nil,
+	}
+	msg, err := GetResponse(c, req, VALUE)
+	if err != nil {
+		return
+	}
+
+	if result, ok := msg.(*ValueResult); ok {
+		value = result.Data
+	}
+
+	if result, ok := msg.(*ChartResultErr); ok {
+		err = errors.New(result.Error)
+	}
+	return
+}
+
+func (m *ValidationManager) compareTwoTrees() []error {
+	return compareTwoTrees(m.templateTree.root, m.testTree.root, m.testTree.lines, m.skippedKeys)
+}
+
+func stringToReadCloser(s string) io.ReadCloser {
+	return ioutil.NopCloser(strings.NewReader(s))
+}
+
+func readCloserToBuffer(readCloser io.ReadCloser, restoreComments, markLineno bool) ([]byte, []string, error) {
```

Comment:
What is closer?

Reply:
It's an union interface of the Reader and Closer in order to support a general `readCloserToBuffer`, which is  unnecessary for current version because it's alway from a string now(no need to close). In previous version, file handler or network resp.Body are both readCloser.